File: examples/ios-oc/CentrifugoIOS/CentrifugoIOS/ViewController.m
//
//  ViewController.m
//  CentrifugoIOS
//
//  Created by nzlov on 2017/12/8.
//  Copyright © 2017年 nzlov. All rights reserved.
//

#import "ViewController.h"
#import <CommonCrypto/CommonDigest.h>
#import <CommonCrypto/CommonHMAC.h>




@interface ViewController ()
@end

@implementation ViewController
@synthesize label;

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    [self.label  setText: @"Connecting..."];
    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    dispatch_async(queue, ^{
        //创建令牌
        CentrifugeCredentials* creds = CentrifugeNewCredentials(@"42", @"1488055494", @"", [self hmac:@"421488055494" withKey:@"109AF84FWF45AS4S5W8F"]);
        //绑定连接事件
        CentrifugeEventHandler* eventHandler = CentrifugeNewEventHandler();
        [eventHandler onConnect:self];
        [eventHandler onDisconnect:self];
        
        //创建客户端连接
        CentrifugeClient *client = CentrifugeNew(@"ws://localhost:8000/connection/websocket", creds, eventHandler, CentrifugeDefaultConfig());
        //连接服务器
        [client connect:NULL];
        //绑定消息事件
        CentrifugeSubEventHandler*  subEventHandler = CentrifugeNewSubEventHandler();
        [subEventHandler onMessage:self];
        //订阅通道
        [client subscribe:@"public:chat" events:subEventHandler error:NULL];
    });
}

- (NSString *)hmac:(NSString *)plaintext withKey:(NSString *)key
{
    const char *cKey  = [key cStringUsingEncoding:NSUTF8StringEncoding];
    const char *cData = [plaintext cStringUsingEncoding:NSUTF8StringEncoding];
    unsigned char cHMAC[CC_SHA256_DIGEST_LENGTH];
    CCHmac(kCCHmacAlgSHA256, cKey, strlen(cKey), cData, strlen(cData), cHMAC);
    NSData *HMACData = [NSData dataWithBytes:cHMAC length:sizeof(cHMAC)];
    const unsigned char *buffer = (const unsigned char *)[HMACData bytes];
    NSMutableString *HMAC = [NSMutableString stringWithCapacity:HMACData.length * 2];
    for (int i = 0; i < HMACData.length; ++i){
        [HMAC appendFormat:@"%02x", buffer[i]];
    }
    
    return HMAC;
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (void)onConnect:(CentrifugeClient *)p0 p1:(CentrifugeConnectContext *)p1 {
    dispatch_async( dispatch_get_main_queue(), ^{
        NSLog(@"Connected");
        [self.label setText:@"Connected"];
    });
}

- (void)onDisconnect:(CentrifugeClient *)p0 p1:(CentrifugeDisconnectContext *)p1 {
    
    dispatch_async( dispatch_get_main_queue(), ^{
        NSLog(@"Disconnect");
        [self.label setText:@"Disconnect"];
    });
}

- (void)onMessage:(CentrifugeSub *)p0 p1:(CentrifugeMessage *)p1 {
    NSLog(@"New Message:%@",p1.data);
    //获得消息后通知服务器已读
    BOOL a = NO;
    [p0 readMessage:p1.uid ret0_:&a error:NULL];
    
    dispatch_async( dispatch_get_main_queue(), ^{
        [self.label setText:p1.data];
    });
}


@end

File: examples/ios-swift/CentrifugoIOS/CentrifugoIOS/ViewController.swift
//
//  ViewController.swift
//  CentrifugoIOS
//
//  Created by Alexander Emelin on 25/02/2017.
//  Copyright © 2017 Alexander Emelin. All rights reserved.
//

import UIKit
import Centrifuge

class ConnectHandler : NSObject, CentrifugeConnectHandlerProtocol {
    var l: UILabel!
    
    func setLabel(l: UILabel!) {
        self.l = l
    }
    
    func onConnect(_ p0: CentrifugeClient!, p1: CentrifugeConnectContext!) {
        DispatchQueue.main.async{
            self.l.text = "Connected";
        }
    }
}

class DisconnectHandler : NSObject, CentrifugeDisconnectHandlerProtocol {
    var l: UILabel!
    
    func setLabel(l: UILabel!) {
        self.l = l
    }
    
    func onDisconnect(_ p0: CentrifugeClient!, p1: CentrifugeDisconnectContext!) {
        DispatchQueue.main.async{
            self.l.text = "Disconnected";
        }
    }
}

class MessageHandler : NSObject, CentrifugeMessageHandlerProtocol,CentrifugeReadHandlerProtocol {
    var l: UILabel!
    
    func setLabel(l: UILabel!) {
        self.l = l
    }
    
    func onMessage(_ p0: CentrifugeSub!, p1: CentrifugeMessage!) {
        DispatchQueue.main.async{
            self.l.text = p1.data()
        }
        do {
            var ok = UnsafeMutablePointer<ObjCBool>.allocate(capacity: 1)
            try  p0.readMessage(p1.uid(), ret0_: ok)
        } catch {
            return
        }
    }
    func onRead(_ p0: CentrifugeSub!, p1: String!) {
        print("Read"+p1)
    }
}

class ViewController: UIViewController {
    
    @IBOutlet weak var label: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        label.text = "Connecting..."
        
        DispatchQueue.main.async{
            let creds = CentrifugeNewCredentials(
                "42","ios_merchant", "1488055494", "", "17445e62d61dfd1fd9e81d0aede358bdeb490e3e9c6cd92f3fd661b72c95b37b"
            )
            
            let eventHandler = CentrifugeNewEventHandler()
            let connectHandler = ConnectHandler()
            connectHandler.setLabel(l: self.label)
            let disconnectHandler = DisconnectHandler()
            disconnectHandler.setLabel(l: self.label)
            
            eventHandler?.onConnect(connectHandler)
            eventHandler?.onDisconnect(disconnectHandler)
            
            
            let url = "ws://192.168.1.200:8000/connection/websocket"
            let client = CentrifugeNew(url, creds, eventHandler, CentrifugeDefaultConfig())
            
            do {
                try client?.connect()
            } catch {
                self.label.text = "Error on connect..."
                return
            }
            
            self.label.text = "Connected"
            
            let subEventHandler = CentrifugeNewSubEventHandler()
            let messageHandler = MessageHandler()
            messageHandler.setLabel(l: self.label)
            subEventHandler?.onMessage(messageHandler)
            subEventHandler?.onRead(messageHandler)
            
            var sub: CentrifugeSub!
            do {
                sub = try client?.subscribe("public:chat", events: subEventHandler)
            } catch {
                DispatchQueue.main.async{
                    self.label.text = "Subscribe error"
                }
                return
            }
            
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
}
